File: src/core/query.ts
import type { ChatMessage } from "../api/types";
import { OpenRouterClient } from "../api/client";
import { runMessageLoop, type MessageLoopCallbacks } from "./message_loop";

export class Conversation {
  private messages: ChatMessage[] = [];
  private client: OpenRouterClient;
  private model: string;

  constructor(client: OpenRouterClient, model: string) {
    this.client = client;
    this.model = model;
  }

  async send(
    userMessage: string,
    callbacks: MessageLoopCallbacks,
  ): Promise<void> {
    this.messages.push({
      role: "user",
      content: userMessage,
    });

    this.messages = await runMessageLoop(
      this.client,
      this.messages,
      this.model,
      callbacks,
    );
  }

  getMessages(): ChatMessage[] {
    return [...this.messages];
  }
}

File: internal/core/query.go
package core

import (
	"context"

	groq "github.com/conneroisu/groq-go"
)

type Conversation struct {
	messages []groq.ChatCompletionMessage
	client   *groq.Client
	model    groq.ChatModel
}

func NewConversation(client *groq.Client, model groq.ChatModel) *Conversation {
	return &Conversation{
		client: client,
		model:  model,
	}
}

func (c *Conversation) Send(ctx context.Context, userMessage string, callbacks MessageLoopCallbacks) {
	c.messages = append(c.messages, groq.ChatCompletionMessage{
		Role:    groq.RoleUser,
		Content: userMessage,
	})
	c.messages = RunMessageLoop(ctx, c.client, c.messages, c.model, callbacks)
}

func (c *Conversation) Messages() []groq.ChatCompletionMessage {
	messages := make([]groq.ChatCompletionMessage, len(c.messages))
	copy(messages, c.messages)
	return messages
}
